src/test/java/org/jenkinsci/plugins/googleplayandroidpublisher/ApkPublisherTest.java

```diff
@@ -0,0 +1,145 @@
+package org.jenkinsci.plugins.googleplayandroidpublisher;
+
+import com.google.api.services.androidpublisher.AndroidPublisher;
+import hudson.FilePath;
+import hudson.model.FreeStyleBuild;
+import hudson.model.FreeStyleProject;
+import hudson.model.Result;
+import hudson.model.queue.QueueTaskFuture;
+import java.io.File;
+import java.util.Arrays;
+import net.dongliu.apk.parser.bean.ApkMeta;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.AndroidUtil;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.JenkinsUtil;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.TestHttpTransport;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.TestUtilImpl;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.TestsHelper;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.responses.FakeAssignTrackResponse;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.responses.FakeCommitResponse;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.responses.FakeListApksResponse;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.responses.FakePostEditsResponse;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.responses.FakePutApkResponse;
+import org.jenkinsci.plugins.googleplayandroidpublisher.internal.responses.FakeUploadApkResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import static hudson.Util.join;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class ApkPublisherTest {
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private AndroidUtil mockAndroid = mock(AndroidUtil.class);
+    private JenkinsUtil jenkinsUtil = spy(TestUtilImpl.class);
+
+    private TestHttpTransport transport = new TestHttpTransport();
+
+    @Before
+    public void setUp() throws Exception {
+        ApkMeta mockMetadata = mock(ApkMeta.class);
+        when(mockMetadata.getPackageName()).thenReturn("org.jenkins.appId");
+        when(mockMetadata.getVersionCode()).thenReturn((long) 42);
+        when(mockMetadata.getMinSdkVersion()).thenReturn("16");
```

Comment:
Minor point: this doesn't really need to be mocked, as this would do the same:
```java
ApkMeta apkMetadata = ApkMeta.newBuilder()
        .setPackageName("org.jenkins.appId")
        .setVersionCode(42L)
        .setMinSdkVersion("16")
        .build();
```

But also I'm not super clear on why we have some of the mocking stuff here, like `mockAndroid`, when I think we can use `TestUtilImpl` (or an anonymous implementation of `AndroidUtil`)?

Reply:
You're right -- Done